RecoMuon/MuonRechitClusterProducer/plugins/CSCRechitClusterProducer.cc

```diff
@@ -0,0 +1,237 @@
+
+#include "fastjet/JetDefinition.hh"
+#include "fastjet/ClusterSequence.hh"
+#include "fastjet/PseudoJet.hh"
+//
+// class declaration
+//
+typedef edm::Ref<CSCRecHit2DCollection> rechitRef;
+typedef std::vector<rechitRef> rechits;
+
+class CSCRechitClusterProducer : public edm::stream::EDProducer<> {
+public:
+  explicit CSCRechitClusterProducer(const edm::ParameterSet&);
+  ~CSCRechitClusterProducer() override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+  edm::EDGetTokenT<CSCRecHit2DCollection> cscRechitInputToken_;
+  const edm::ESGetToken<CSCGeometry, MuonGeometryRecord> m_cscGeometryToken;
+  typedef std::vector<reco::MuonCSCRecHitCluster> MuonCSCRecHitClusterCollection;
+
+  // ----------member data ---------------------------
+protected:
+  double rParam_;   // distance paramter
+  int nRechitMin_;  // min number of rechits
+  std::vector<rechitRef> inputs_;
+  std::vector<rechitRef> getConstituents(const std::vector<fastjet::PseudoJet>& fjConstituents);
+};
+
+//
+// constants, enums and typedefs
+//
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+CSCRechitClusterProducer::CSCRechitClusterProducer(const edm::ParameterSet& iConfig)
+    : m_cscGeometryToken(esConsumes<CSCGeometry, MuonGeometryRecord>()) {
+  rParam_ = iConfig.getParameter<double>("rParam");
+  nRechitMin_ = iConfig.getParameter<int>("nRechitMin");
+
+  cscRechitInputToken_ = consumes<CSCRecHit2DCollection>(iConfig.getParameter<edm::InputTag>("cscRecHitLabel"));
+  produces<MuonCSCRecHitClusterCollection>();
+}
+
+CSCRechitClusterProducer::~CSCRechitClusterProducer() {}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void CSCRechitClusterProducer::produce(edm::Event& ev, const edm::EventSetup& iSetup) {
+  edm::ESHandle<CSCGeometry> cscG;
+  cscG = iSetup.getHandle(m_cscGeometryToken);
+
+  edm::Handle<CSCRecHit2DCollection> cscRechits;
+
+  ev.getByToken(cscRechitInputToken_, cscRechits);
+
+  std::vector<fastjet::PseudoJet> fjInput;
+
+  fastjet::JetDefinition jet_def(fastjet::cambridge_algorithm, rParam_);
+
+  CSCRecHit2DCollection::const_iterator recIt;
+
+  inputs_.clear();
+  fjInput.clear();
+
+  for (recIt = cscRechits->begin(); recIt != cscRechits->end(); recIt++) {
+    auto cscRechit = (*recIt);
+    LocalPoint cscRecHitLocalPosition = cscRechit.localPosition();
+    CSCDetId cscdetid = cscRechit.cscDetId();
+    const CSCChamber* cscchamber = cscG->chamber(cscdetid);
+    if (cscchamber) {
+      GlobalPoint globalPosition = cscchamber->toGlobal(cscRecHitLocalPosition);
+      float x = globalPosition.x();
+      float y = globalPosition.y();
+      float z = globalPosition.z();
+      rechitRef csc_ref = rechitRef(cscRechits, recIt - cscRechits->begin());
+      inputs_.push_back(csc_ref);
+      fjInput.push_back(fastjet::PseudoJet(x, y, z, 1.0));
```

Comment:
Yes, the mass is negative, but it's not used during the clustering process.
Here's an example output:
```
 unit     mag: x = -84.8177  y = -407.462 z = 837.025 mass = -270154 e = 289
 global  mag: x = -84.8177  y = -407.462 z = 837.025 mass = 12893.1 e = 270461
```

Reply:
I saw some weird issues when the mass was negative (though they only appeared if I stored the fastjet in the event as a BasicJet). IMHO I think it would be a good idea to follow @slava77 suggestion unless there's a strong reason not to.

Reply:
I don't have a strong preference, since it does not change the end result.  

@slava77 I tested with `sqrt(globalPosition.mag())` but that still yields negative mass for some reason.
The above result is obtained with
```
fjInput.push_back(fastjet::PseudoJet(x, y, z, std::sqrt(x*x+y*y+z*z)));
```
can I use this instead?

Reply:
sorry, no sqrt, just `globalPosition.mag()`.

Reply:
btw use
```fjInput.emplace_back(x, y, z, std::sqrt(x*x+y*y+z*z))```  (or .mag() that computes the very same sqrt)

Reply:
Done here:
https://github.com/cms-sw/cmssw/blob/86b6c2652ceb7ea2896c5c4774d7150e8a8e4c40/RecoMuon/MuonRechitClusterProducer/plugins/RechitClusterProducer.cc#L101